tour_model_eval/trips_in_bins_and_clusters.ipynb

```diff
@@ -42,293 +49,789 @@
+    "Note: should run \"Trips from bins above the cutoff only\" first"
    ]
   },
   {
    "cell_type": "code",
    "execution_count": null,
-   "id": "ultimate-bottom",
+   "id": "lightweight-plaintiff",
+   "metadata": {
+    "scrolled": false
+   },
+   "outputs": [],
+   "source": [
+    "# copied from the cluster pipeline code so we can work with the featurization code directly\n",
+    "# min clusters set to 0 instead of len(bins), to compare with the bins\n",
+    "feat = featurization.featurization(bin_trips)\n",
+    "min = 0\n",
+    "max = int(math.ceil(1.5 * len(bins)))\n",
+    "feat.cluster(min_clusters=min, max_clusters=max)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "technical-victoria",
    "metadata": {},
    "outputs": [],
    "source": [
-    "sim = similarity.similarity(trips, radius)\n",
-    "sim.bin_data()"
+    "logging.debug('number of clusters: %d' % feat.clusters)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "young-pottery",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "#labels have to be in order in the colormap index\n",
+    "labels_clt = feat.labels\n",
+    "labels_clt = list(set(sorted(labels_clt)))\n",
+    "labels_clt"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "western-fourth",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "cmp_clusters_cutoff = cm.linear.Set1_07.to_step((feat.clusters),index=[i for i in labels_clt])\n",
+    "cmp_clusters_cutoff"
    ]
   },
   {
    "cell_type": "code",
    "execution_count": null,
-   "id": "scheduled-scout",
+   "id": "undefined-society",
    "metadata": {
     "scrolled": true
    },
    "outputs": [],
    "source": [
-    "# Plot all the bin trips on the map, later we should use different color for different bins, but not sure whether it works.\n",
-    "m = folium.Map(location=[trips[0].data.start_loc[\"coordinates\"][1], trips[0].data.start_loc[\"coordinates\"][0]],zoom_start=12, max_zoom= 30, control_scale=True)\n",
-    "t_index = 0\n",
-    "for bin in sim.bins:    \n",
-    "    for i in range (len(bin)):\n",
-    "        layer = folium.PolyLine([[trips[t_index].data.start_loc[\"coordinates\"][1],trips[t_index].data.start_loc[\"coordinates\"][0]],\n",
-    "                                [trips[t_index].data.end_loc[\"coordinates\"][1],trips[t_index].data.end_loc[\"coordinates\"][0]]],weight=2, color='orange')\n",
-    "        t_index += 1\n",
-    "        layer.add_to(m)\n",
-    "        \n",
-    "m"
+    "#Plot the clusters above the cutoff only with different colors,to see if the algorithm wrongly puts different trips into the same cluster\n",
+    "cluster_cutoff = folium.Map(location=[feat.points[0][1], feat.points[0][0]],zoom_start=12, max_zoom= 30, control_scale=True)\n",
+    "\n",
+    "print(feat.labels)\n",
+    "if feat.labels:             \n",
+    "    for i in range(len(feat.points)):\n",
```

Comment:
again, can use `for i, point in enumerate(feat.points)` or even `for color, point in zip(color, points)`